core/include/vecmem/memory/device_atomic_ref.hpp

```diff
@@ -22,11 +29,15 @@ namespace vecmem {
 using memory_order = ::sycl::memory_order;
 
 /// @c vecmem::atomic_ref equals @c sycl::atomic_ref with "modern SYCL"
-template <typename T>
+template <typename T, address_space address = address_space::global>
 using device_atomic_ref =
     ::sycl::atomic_ref<T, ::sycl::memory_order::relaxed,
-                       ::sycl::memory_scope::device,
-                       ::sycl::access::address_space::global_space>;
+                       address == address_space::global
+                           ? ::sycl::memory_scope::device
+                           : ::sycl::memory_scope::work_group,
+                       address == address_space::global
+                           ? ::sycl::access::address_space::global_space
+                           : ::sycl::access::address_space::local_space>;
```

Comment:
:thinking: Let's do it more properly. It's very likely that 2 values is what `vecmem::device_address_space` is ever going to have. But even so, let's not hard-code this assumption everywhere in our code.

Instead, let's use some "trait classes" here. Like:

```c++
template <device_address_space address>
struct sycl_address {};
template <>
struct sycl_address<device_address_space::global> {
   using memory_scope = ::sycl::memory_scope::device;
   using address_space = ::sycl::access::address_space::global_space;
};
template <>
struct sycl_address<device_address_space::local> {
   using memory_scope = ::sycl::memory_scope::work_group;
   using address_space = ::sycl::access::address_space::local_space;
};
template <typename T, device_address_space address = device_address_space::global>
using device_atomic_ref = ::sycl::atomic_ref<T, sycl_address<address>::memory_scope, sycl_address<address>::address_space>;
```

(Code of course not actually tested...)

Reply:
And to make it clean, the actual struct should have some nicer name. Something like `vecmem::details::sycl_address_space`. (Matching with the `vecmem::device_address_space` name...)